File: 1207-delete-nodes-and-return-forest/delete-nodes-and-return-forest.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        Set<Integer> toDelete = new HashSet<>();
        for (int val : to_delete) {
            toDelete.add(val);
        }
        
        List<TreeNode> forest = new ArrayList<>();
        
        root = deleteNodes(root, toDelete, forest);
        
        if (root != null) {
            forest.add(root);
        }
        
        return forest;
    }
    
    private TreeNode deleteNodes(TreeNode node, Set<Integer> toDelete, List<TreeNode> forest) {
        if (node == null) {
            return null;
        }
        
        boolean deleted = toDelete.contains(node.val);
        
        node.left = deleteNodes(node.left, toDelete, forest);
        node.right = deleteNodes(node.right, toDelete, forest);
        
        if (deleted) {
            if (node.left != null) {
                forest.add(node.left);
            }
            if (node.right != null) {
                forest.add(node.right);
            }
            return null;
        }
        
        return node;
    }
}

File: 1207-delete-nodes-and-return-forest/delete-nodes-and-return-forest.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

struct TreeNode* deleteNodes(struct TreeNode* node, bool* to_delete_set, struct TreeNode** forest, int* returnSize);

struct TreeNode** delNodes(struct TreeNode* root, int* to_delete, int to_deleteSize, int* returnSize) {
    bool* to_delete_set = calloc(1001, sizeof(bool)); // Assuming values are in range 1-1000
    for (int i = 0; i < to_deleteSize; i++) {
        to_delete_set[to_delete[i]] = true;
    }
    
    struct TreeNode** forest = malloc(1001 * sizeof(struct TreeNode*));
    *returnSize = 0;
    
    root = deleteNodes(root, to_delete_set, forest, returnSize);
    
    if (root != NULL) {
        forest[(*returnSize)++] = root;
    }
    
    free(to_delete_set);
    return forest;
}

struct TreeNode* deleteNodes(struct TreeNode* node, bool* to_delete_set, struct TreeNode** forest, int* returnSize) {
    if (node == NULL) return NULL;
    
    bool should_delete = to_delete_set[node->val];
    
    node->left = deleteNodes(node->left, to_delete_set, forest, returnSize);
    node->right = deleteNodes(node->right, to_delete_set, forest, returnSize);
    
    if (should_delete) {
        if (node->left != NULL) {
            forest[(*returnSize)++] = node->left;
        }
        if (node->right != NULL) {
            forest[(*returnSize)++] = node->right;
        }
        return NULL;
    }
    
    return node;
}